components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -222,13 +234,11 @@ abstract class AbstractFetchDownloadService : Service() {
      * notifications are not dropped
      */
     @Suppress("ComplexMethod")
-    private fun updateDownloadNotification() {
-        for (download in downloadJobs.values) {
-            if (download.notifiedStopped) { continue }
-
+    private fun updateDownloadNotification(download: DownloadJobState) {
             // Dispatch the corresponding notification based on the current status
             val notification = when (getDownloadJobStatus(download)) {
                 DownloadJobStatus.ACTIVE -> {
+                    download.uiStatus = DownloadJobStatus.ACTIVE
```

Comment:
I'm a little confused about `uiStatus` here. Is it just the same as `status`? Is there a way we can simplify our logic so we're not having to book-keep on two similarly named variables (perhaps using a computed property or using just one status)?

Reply:
I had to add an extra property to make sure we don't loose any UI updates.  I noticed in some cases we were missing some notifications updates as the notifications were not reflecting the real status in the UI.

For example:

We have [this condition](https://github.com/mozilla-mobile/android-components/blob/8bba0959d5527da80f1c4ca3117ccbf875bc251f/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L227), that doesn't allow to update the notification if the `notifiedStopped` is `true`
https://github.com/mozilla-mobile/android-components/blob/8bba0959d5527da80f1c4ca3117ccbf875bc251f/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L227

As `updateDownloadNotification` [could run in parallel as user interactions](https://github.com/mozilla-mobile/android-components/blob/8bba0959d5527da80f1c4ca3117ccbf875bc251f/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L210), in some cases we were updating [notifiedStopped to true](https://github.com/mozilla-mobile/android-components/blob/8bba0959d5527da80f1c4ca3117ccbf875bc251f/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L386), without making sure this update was posted to the UI. Causing the UI to not be in sync with the data, and making the notification not responsive as the `notifiedStopped`  was `true`, [we didn't allow updates on the notification](https://github.com/mozilla-mobile/android-components/blob/8bba0959d5527da80f1c4ca3117ccbf875bc251f/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L227).